### PR TITLE
Lseek data and hole

### PR DESCRIPTION
**Pull request by GBuella**
This change is [Reviewable](https://reviewable.io/reviews/marcinslusarz/pmemfile/91)

### CODE REVIEW

File: src/libpmemfile-posix/data.c [hunk 1]
```diff
@@ -785,6 +785,98 @@ pmemfile_read(PMEMfilepool *pfp, PMEMfile *file, void *buf, size_t count)
...
+		else
+			return (off64_t)vinode->first_block->offset;
+	}
+
+	if (is_offset_in_block(block, (uint64_t)offset))
+		return offset;
+
+	block = D_RW(block->next);
+
+	if (block == NULL || block->offset > (uint64_t)offset)
```
Comment (marcinslusarz) [inline src/libpmemfile-posix/data.c]: 2nd condition seems to be wrong.

File: src/libpmemfile-posix/data.c [hunk 2]
```diff
@@ -785,6 +785,98 @@ pmemfile_read(PMEMfilepool *pfp, PMEMfile *file, void *buf, size_t count)
...
+	if (offset < 0)
+		offset = 0;
+
+	if (offset > fsize) {
+		*new_errno = ENXIO;
+		return -1;
+	}
+
+	if (!vinode_is_regular_file(vinode))
+		return fsize; /* seeking in directory */
```
Comment (marcinslusarz) [inline src/libpmemfile-posix/data.c]: This is equivalent to SEEK_END, which is forbidden for directories.

File: src/libpmemfile-posix/data.c [hunk 3]
```diff
@@ -785,6 +785,98 @@ pmemfile_read(PMEMfilepool *pfp, PMEMfile *file, void *buf, size_t count)
...
+		struct pmemfile_block *next = D_RW(block->next);
+
+		if (block_end >= offset)
+			offset = block_end; /* seek to the end of block */
+
+		if (next == NULL)
+			break; /* the rest of the file is treated as a hole */
+
+		if (next != NULL && offset < (off64_t)next->offset)
+			break; /* offset is in a hole between two blocks */
```
Comment (marcinslusarz) [inline src/libpmemfile-posix/data.c]: "next != NULL" is not needed - you just checked that

Comment (marcinslusarz) [inline src/libpmemfile-posix/data.c]: Also, you need to set offset to next->offset.

File: src/libpmemfile-posix/data.c [hunk 4]
```diff
@@ -785,6 +785,98 @@ pmemfile_read(PMEMfilepool *pfp, PMEMfile *file, void *buf, size_t count)
...
+
+	if (!vinode_is_regular_file(vinode))
+		return fsize; /* seeking in directory */
+
+	if (vinode->blocks == NULL)
+		vinode_rebuild_block_tree(vinode);
+
+	struct pmemfile_block *block = find_block(vinode, (uint64_t)offset);
+
+	while (block != NULL && offset < fsize) {
```
Comment (marcinslusarz) [inline src/libpmemfile-posix/data.c]: Why do you need this to be a loop? Hole will be after current block.

File: tests/posix/rw/rw.cpp [hunk 5]
```diff
@@ -978,29 +978,82 @@ TEST_F(rw, o_append)
...
+	 * Seeking to hole, or to data should fail with offset
+	 * greater than the file size.
+	 */
+	errno = 0;
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 1, PMEMFILE_SEEK_HOLE), -1);
+	ASSERT_EQ(errno, ENXIO);
+	errno = 0;
+	ASSERT_EQ(pmemfile_lseek(pfp, f, 1, PMEMFILE_SEEK_DATA), -1);
+	ASSERT_EQ(errno, ENXIO);
+	errno = 0;
```
Comment (marcinslusarz) [inline tests/posix/rw/rw.cpp]: not needed

File: tests/posix/rw/rw.cpp [hunk 6]
```diff
@@ -1011,7 +1064,120 @@ TEST_F(rw, sparse_files)
...
+	 * that can achieve this is fallocate.
+	 */
+	size = 0x40000;
+	hole = size / 2; /* The hole starts at this offset */
+	r = pmemfile_ftruncate(pfp, f, size);
+	ASSERT_EQ(r, 0) << strerror(errno);
+	r = pmemfile_fallocate(pfp, f,
+			       PMEMFILE_FL_PUNCH_HOLE | PMEMFILE_FL_KEEP_SIZE,
+			       hole, size);
+	ASSERT_EQ(r, 0) << strerror(errno);
```
Comment (marcinslusarz) [inline tests/posix/rw/rw.cpp]: FTR, the code below assumes that ftruncate didn't cheat and allocated all blocks. Maybe it's worth enforcing or at least adding a comment.

**Comment by GBuella** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/libpmemfile-posix/data.c, line 824 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91#-Kh7PWFvjWwMhqbjmiUI-r1-824:-Kh7uDqIEXOiGcjTgF-f:b-z3vp35) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/99c60a22f862d268338dbc1acad15e529873927f/src/libpmemfile-posix/data.c#L824)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

2nd condition seems to be wrong.
</blockquote></details>

It is wrong indeed, I mean to check ```block->offset > fsize```

---

*[src/libpmemfile-posix/data.c, line 856 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91#-Kh7PWFvjWwMhqbjmiUI-r1-856:-Kh7thMmTT5L0Fh1kU_z:bf5usyd) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/99c60a22f862d268338dbc1acad15e529873927f/src/libpmemfile-posix/data.c#L856)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Why do you need this to be a loop? Hole will be after current block.
</blockquote></details>

What do you mean? There might a lot a of blocks, without any holes between them.

---

*[src/libpmemfile-posix/data.c, line 868 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91#-Kh7PWFvjWwMhqbjmiUI-r1-868:-Kh7u8-S-31OQ4kA042O:bwvr57c) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/99c60a22f862d268338dbc1acad15e529873927f/src/libpmemfile-posix/data.c#L868)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Also, you need to set offset to next->offset.
</blockquote></details>

Why would I need to set offset to next->offset?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91)*

File: src/libpmemfile-posix/data.c [hunk 7]
```diff
@@ -785,6 +785,98 @@ pmemfile_read(PMEMfilepool *pfp, PMEMfile *file, void *buf, size_t count)
...
+
+	off64_t fsize = (off64_t)vinode->inode->size;
+
+	if (offset > fsize) {
+		*new_errno = ENXIO;
+		return -1;
+	}
+
+	if (!vinode_is_regular_file(vinode))
+		return offset;
```
Comment (marcinslusarz) [inline src/libpmemfile-posix/data.c]: This check needs to be above file size check.

**Comment by marcinslusarz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/libpmemfile-posix/data.c, line 856 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91#-Kh7PWFvjWwMhqbjmiUI-r1-856:-Kh7xwmjLr2LPr-zNZeg:byr4flj) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/99c60a22f862d268338dbc1acad15e529873927f/src/libpmemfile-posix/data.c#L856)):*
<details><summary><i>Previously, GBuella (Gábor Buella) wrote…</i></summary><blockquote>

What do you mean? There might a lot a of blocks, without any holes between them.
</blockquote></details>

Right.

---

*[src/libpmemfile-posix/data.c, line 868 at r1](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91#-Kh7PWFvjWwMhqbjmiUI-r1-868:-Kh7yb4-K74FqvdXlHpA:b-b5f8en) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/99c60a22f862d268338dbc1acad15e529873927f/src/libpmemfile-posix/data.c#L868)):*
<details><summary><i>Previously, GBuella (Gábor Buella) wrote…</i></summary><blockquote>

Why would I need to set offset to next->offset?
</blockquote></details>

Argh, right. I was thinking about SEEK_DATA instead of SEEK_HOLE.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91)*

**Comment by marcinslusarz** [comment]
*[src/libpmemfile-posix/data.c, line 825 at r2](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91#-Kh8JDfq7rqP3M_PY62E:-Kh8JDfq7rqP3M_PY62F:b-648ufm) ([raw file](https://github.com/marcinslusarz/pmemfile/blob/21882cdcfa296e60baa0780fde9eeb65b31db69d/src/libpmemfile-posix/data.c#L825)):*
> ```C
> {
> 	if (vinode->blocks == NULL)
> 		vinode_rebuild_block_tree(vinode);
> ```

Please move this code to lseek64_seek_data_or_hole in the next patch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91)*

**Comment by marcinslusarz** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/marcinslusarz/pmemfile/91#-:-Kh8JgfdyflleHvhq2xg:bnfp4nl)*

**Pull request merged**